File: sol/solution/0001-0100/0018/Solution.scala
object Solution {
    def fourSum(nums: Array[Int], target: Int): List[List[Int]] = {
        val result = scala.collection.mutable.ListBuffer[List[Int]]()
        val n = nums.length
        if (n < 4) return result.toList
        val sorted = nums.sorted
        for (i <- 0 until n - 3) {
            if (i > 0 && sorted(i) == sorted(i - 1)) {}
            else {
                for (j <- i + 1 until n - 2) {
                    if (j > i + 1 && sorted(j) == sorted(j - 1)) {}
                    else {
                        var left = j + 1
                        var right = n - 1
                        while (left < right) {
                            val sum = sorted(i).toLong + sorted(j) + sorted(left) + sorted(right)
                            if (sum == target) {
                                result += List(sorted(i), sorted(j), sorted(left), sorted(right))
                                while (left < right && sorted(left) == sorted(left + 1)) left += 1
                                while (left < right && sorted(right) == sorted(right - 1)) right -= 1
                                left += 1
                                right -= 1
                            } else if (sum < target) left += 1
                            else right -= 1
                        }
                    }
                }
            }
        }
        result.toList
    }
}

File: sol/solution/0001-0100/0018/Solution.c
int cmp(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}
int** fourSum(int* nums, int numsSize, int target, int* returnSize, int** returnColumnSizes) {
    *returnSize = 0;
    if (numsSize < 4) return NULL;
    qsort(nums, numsSize, sizeof(int), cmp);
    int capacity = 256;
    int** result = malloc(capacity * sizeof(int*));
    *returnColumnSizes = malloc(capacity * sizeof(int));
    for (int i = 0; i < numsSize - 3; i++) {
        if (i > 0 && nums[i] == nums[i - 1]) continue;
        long long min1 = (long long)nums[i] + nums[i + 1] + nums[i + 2] + nums[i + 3];
        if (min1 > target) break;
        long long max1 = (long long)nums[i] + nums[numsSize - 1] + nums[numsSize - 2] + nums[numsSize - 3];
        if (max1 < target) continue;
        for (int j = i + 1; j < numsSize - 2; j++) {
            if (j > i + 1 && nums[j] == nums[j - 1]) continue;
            long long min2 = (long long)nums[i] + nums[j] + nums[j + 1] + nums[j + 2];
            if (min2 > target) break;
            long long max2 = (long long)nums[i] + nums[j] + nums[numsSize - 1] + nums[numsSize - 2];
            if (max2 < target) continue;
            int left = j + 1, right = numsSize - 1;
            while (left < right) {
                long long sum = (long long)nums[i] + nums[j] + nums[left] + nums[right];
                if (sum == target) {
                    if (*returnSize == capacity) {
                        capacity *= 2;
                        result = realloc(result, capacity * sizeof(int*));
                        *returnColumnSizes = realloc(*returnColumnSizes, capacity * sizeof(int));
                    }
                    int* quad = result[*returnSize] = malloc(4 * sizeof(int));
                    quad[0] = nums[i];
                    quad[1] = nums[j];
                    quad[2] = nums[left];
                    quad[3] = nums[right];
                    (*returnColumnSizes)[*returnSize] = 4;
                    (*returnSize)++;
                    int last_left = nums[left], last_right = nums[right];
                    while (left < right && nums[left] == last_left) left++;
                    while (left < right && nums[right] == last_right) right--;
                } else if (sum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }
    }
    if (*returnSize == 0) {
        free(result);
        free(*returnColumnSizes);
        return NULL;
    }
    return result;
}
